toolchain/lc-compile/src/literal.c

```diff
@@ -87,6 +161,155 @@ int MakeIntegerLiteral(const char *p_token, long *r_literal)
     return 1;
 }
 
+// Match 0 | [1-9][0-9]*
+static int
+__MatchInt(const char **x_token_ptr, int* r_found)
+{
+	// If there are no digits, there is no int.
+	if (!isdigit(*(*x_token_ptr)))
+	{
+		*r_found = 0;
+		return 1;
+	}
+	
+	// If there is zero before a digit, it is malformed.
+	if ((*x_token_ptr)[0] == '0' &&
+		isdigit((*x_token_ptr)[1]))
+	{
+		return 1;
+	}
+	
+	*x_token_ptr += 1;
+	
+	// Match [0-9]*
+	while(isdigit(*x_token_ptr[0]))
+	{
+		*x_token_ptr += 1;
+	}
+	
+	*r_found = 1;
+	
+	return 1;
+}
+
+// Match [0-9]+
+static int
+__MatchZeroesInt(const char **x_token_ptr, int* r_found)
+{
+	// If there are no digits, there is no int.
+	if (!isdigit(*x_token_ptr[0]))
+	{
+		*r_found = 0;
+		return 1;
+	}
+	
+	// Match [0-9]*
+	while(isdigit(*x_token_ptr[0]))
+	{
+		*x_token_ptr += 1;
+	}
+	
+	*r_found = 1;
+	
+	return 1;
+}
+
+int IsDoubleLiteral(const char *p_token)
+{
+	// We check the pattern against:
+	//   I:Int P:Period F:ZeroesInt E:(Exp PlusMinus Int)
+	// Where the valid combinations are:
+	//   I+P
+	//   I+E
+	//   I+P+E
+	//   I+P+F
+	//   I+P+F+E
+	//   P+F
+	//   P+F+E
+	
+	int t_found_int = 0;
+	int t_found_point = 0;
+	int t_found_frac = 0;
+	int t_found_exp = 0;
+	int t_found_bits = 0;
+	const char *t_token_ptr = p_token;
+
+	// Match I - if the Int is malformed, its not a double.
+	if (!__MatchInt(&t_token_ptr, &t_found_int))
+	{
+		return 0;
+	}
+	
+	// Match P
+	if (*t_token_ptr == '.')
+	{
+		t_found_point = 1;
+		t_token_ptr++;
+	}
+	
+	// Match F - if the ZeroesInt is malformed, its not a double
+	if (!__MatchZeroesInt(&t_token_ptr, &t_found_frac))
+	{
+		return 0;
+	}
+	
+	// Match E
+	if (*t_token_ptr == 'e' || *t_token_ptr == 'E')
+	{
+		t_token_ptr++;
+		if (*t_token_ptr == '+' || *t_token_ptr == '-')
+		{
+			t_token_ptr++;
+		}
+		if (!__MatchInt(&t_token_ptr, &t_found_exp))
+		{
+			return 0;
+		}
+	}
+	
+	// If we aren't at end of the token, its not a double.
+	if (*t_token_ptr != '\0')
+	{
+		return 0;
+	}
+	
+	// Now check valid combinations by computing it as a bitset
+	t_found_bits =
+		(t_found_int << 3) |
+		(t_found_point << 2) |
+		(t_found_frac << 1) |
+		(t_found_exp << 0);
+	
+	#define I 8
+	#define P 4
+	#define F 2
+	#define E 1
```

Comment:
Please move these definitions up the function and use them when computing `t_found_bits`, e.g.

```C
t_found_bits =
    (t_found_int ? I : 0) |
    (t_found_point ? P : 0) |
    (t_found_frac ? F : 0) |
    (t_found_exp ? E : 0);
```